modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -182,12 +189,112 @@ public CompletableFuture<Object> invoke(ReplicaRequest request) {
                         return processTxFinishAction((TxFinishReplicaRequest) request);
                     } else if (request instanceof TxCleanupReplicaRequest) {
                         return processTxCleanupAction((TxCleanupReplicaRequest) request);
+                    } else if (request instanceof ReadOnlySingleRowReplicaRequest) {
+                        return processReadOnlySingleEntryAction((ReadOnlySingleRowReplicaRequest) request);
+                    } else if (request instanceof ReadOnlyMultiRowReplicaRequest) {
+                        return processReadOnlyMultiEntryAction((ReadOnlyMultiRowReplicaRequest) request);
+                    } else if (request instanceof ReadOnlyScanRetrieveBatchReplicaRequest) {
+                        return processReadOnlyScanRetrieveBatchAction((ReadOnlyScanRetrieveBatchReplicaRequest) request);
                     } else {
                         throw new UnsupportedReplicaRequestException(request.getClass());
                     }
                 });
     }
 
+    /**
+     * Processes retrieve batch for read only transaction.
+     *
+     * @param request Read only retrieve batch request.
+     * @return Result future.
+     */
+    private CompletableFuture<Object> processReadOnlyScanRetrieveBatchAction(ReadOnlyScanRetrieveBatchReplicaRequest request) {
+        UUID txId = request.transactionId();
+        int batchCount = request.batchSize();
+
+        IgniteUuid cursorId = new IgniteUuid(txId, request.scanId());
+
+        PartitionTimestampCursor timestampCursor = mvDataStorage.scan(row -> true, request.timestamp());
+
+        Cursor<BinaryRow> cursor = cursors.computeIfAbsent(cursorId, id -> new BinaryRowCursorWrapper(timestampCursor));
+
+        ArrayList<BinaryRow> batchRows = new ArrayList<>(batchCount);
+
+        for (int i = 0; i < batchCount && cursor.hasNext(); i++) {
+            batchRows.add(cursor.next());
+        }
+
+        return CompletableFuture.completedFuture(batchRows);
+    }
+
+    /**
+     * Processes single entry request for read only transaction.
+     *
+     * @param request Read only single entry request.
+     * @return Result future.
+     */
+    private CompletableFuture<Object> processReadOnlySingleEntryAction(ReadOnlySingleRowReplicaRequest request) {
+        ByteBuffer searchKey = request.binaryRow().keySlice();
+
+        UUID indexId = indexIdOrDefault(indexScanId/*request.indexToUse()*/);
+
+        if (request.requestType() !=  RequestType.RO_GET) {
+            throw new IgniteInternalException(Replicator.REPLICA_COMMON_ERR,
+                    IgniteStringFormatter.format("Unknown single request [actionType={}]", request.requestType()));
+        }
+
+        //TODO: Use timestamp to find a row id.
+        RowId rowId = rowIdByKey(indexId, searchKey);
```

Comment:
There may be multiple rowIds. 
k1 -> rowId1
tx begin
  remove k1
  insert k1->rowId2
tx commit

As a result there are two k1 entries in index.